src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpoint.java

```diff
@@ -0,0 +1,258 @@
+        return publicKey;
+    }
+
+    /**
+     * Gets the supported protocols if available.
+     *
+     * @return the supported protocols or {@code null}.
+     */
+    @CheckForNull
+    public Set<String> getProtocols() {
+        return protocols;
+    }
+
+    /**
+     * Checks if the named protocol is supported.
+     *
+     * @param name the name of the protocol to check.
+     * @return {@code false} if and only if the endpoint reports supported protocols and the named protocol is not in
+     * the list of supported protocols.
+     */
+    public boolean isProtocolSupported(@Nonnull String name) {
+        return protocols == null || protocols.contains(name);
+    }
+
+    public Socket open(int socketTimeout) throws IOException {
+        boolean isHttpProxy = false;
+        InetSocketAddress targetAddress = null;
+        SocketChannel channel = null;
+        try {
+            targetAddress = JnlpAgentEndpointResolver.getResolvedHttpProxyAddress(host, port);
+
+            if (targetAddress == null) {
+                targetAddress = new InetSocketAddress(host, port);
+            } else {
+                isHttpProxy = true;
+            }
+
+            // We open the socket using SocketChannel so that we are assured that the socket will always have
+            // a socket channel. Sockets opened via Socket.open will typically not have a SocketChannel
+            // and thus we will not have the ability to use NIO if we want to.
+            channel = SocketChannel.open(targetAddress);
+            Socket socket = channel.socket();
+
+            socket.setTcpNoDelay(true); // we'll do buffering by ourselves
+
+            // set read time out to avoid infinite hang. the time out should be long enough so as not
+            // to interfere with normal operation. the main purpose of this is that when the other peer dies
+            // abruptly, we shouldn't hang forever, and at some point we should notice that the connection
+            // is gone.
+            socket.setSoTimeout(socketTimeout);
+
+            if (isHttpProxy) {
+                String connectCommand = String.format("CONNECT %s:%s HTTP/1.1\r\nHost: %s\r\n\r\n", host, port, host);
+                socket.getOutputStream().write(connectCommand.getBytes("UTF-8")); // TODO: internationalized domain names
+
+                BufferedInputStream is = new BufferedInputStream(socket.getInputStream());
+                String line = readLine(is);
+                String[] responseLineParts = line.split(" ");
+                if (responseLineParts.length < 2 || !responseLineParts[1].equals("200"))
+                    throw new IOException("Got a bad response from proxy: " + line);
+                while (!readLine(is).isEmpty()) {
+                    // Do nothing, scrolling through headers returned from proxy
+                }
+            }
+            return socket;
+        } catch (IOException e) {
+            IOUtils.closeQuietly(channel);
```

Comment:
Generally we need our own closeQuietly() implementation, which prints the message to logger at least